src/Streamly/Internal/Data/Fold/Types.hs

```diff
@@ -657,35 +695,56 @@ runStep _ _ = undefined
 --
 {-# INLINE many #-}
 many :: Monad m => Fold m b c -> Fold m a b -> Fold m a c
-many (Fold fstep finitial fextract) (Fold step1 initial1 extract1) =
+many (Fold cstep cinitial cextract) (Fold sstep sinitial sextract) =
     Fold step initial extract
 
     where
 
+    -- cs = collect state
+    -- ss = split state
+
+    loopTillPartial cs sb = do
+        cres <- cstep cs sb
+        case cres of
+            Partial cs1 -> do
+                sres <- sinitial
+                case sres of
+                    Partial ss -> return $ Partial $ Tuple' ss cs
+                    Done sb1 -> loopTillPartial cs1 sb1
+            Done cb -> return $ Done cb
+
     initial = do
-        ps <- initial1
-        fs <- finitial
-        pure (Tuple' ps fs)
+        cres <- cinitial
```

Comment:
I had to rename these because it was hard to keep track of stuff when half asleep.